File: tests/stress/long_running.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
A long running test that performs
random CRUD ops on a bolt database.

Parameterized with vertex and edge counts around which
the graph state oscilates.
"""

import logging
import multiprocessing
import neo4j.exceptions
import random
import time
from collections import defaultdict

import common


log = logging.getLogger(__name__)


INDEX_FORMAT = "indexed_label{}"


def random_element(lst):
    return lst[random.randint(0, len(lst) - 1)]


def bernoulli(p):
    return random.random() < p


class Graph():
    """
    Exposes functions for working on a graph, and tracks some
    statistics about graph state.
    """

    def __init__(self, vertex_count, edge_count, labels=5):
        """
        Args:
            vertex_count - int, desired vertex count
            edge_count - int, desired edge count
            labels - int, the number of labels to use
        """

        # desired vertex and edge counts
        self.vertex_count = vertex_count
        self.edge_count = edge_count

        # storage
        self.edges = []
        self.vertices = []
        self.labels = {"label%d" % i: [] for i in range(labels)}

        # info about query failures, maps exception string representations into
        # occurence counts
        self._query_failure_counts = defaultdict(int)

    def add_query_failure(self, reason):
        self._query_failure_counts[reason] += 1

    def query_failures(self):
        return dict(self._query_failure_counts)


class GraphSession():
    """
    Encapsulates a Graph and a Bolt session and provides CRUD op functions.
    Also defines a run-loop for a generic exectutor, and a graph state
    verification function.
    """

    def __init__(self, sid, graph, session):
        self.sid = sid

        # the label in the database that is indexed
        # used for matching vertices faster
        self.indexed_label = INDEX_FORMAT.format(sid)

        self.vertex_id = 1
        self.edge_id = 1

        self.graph = graph
        self.session = session
        self.executed_queries = 0
        self._start_time = time.time()

    @property
    def v(self):
        return self.graph.vertices

    @property
    def e(self):
        return self.graph.edges

    def execute(self, query):
        log.debug("Runner %d executing query: %s", self.sid, query)
        self.executed_queries += 1
        try:
            return self.session.run(query).data()
        except neo4j.exceptions.ServiceUnavailable as e:
            raise e
        except Exception as e:
            self.graph.add_query_failure(str(e))
            return None

    def create_vertices(self, vertices_count):
        query = ""
        if vertices_count == 0: return
        for _ in range(vertices_count):
            query += "CREATE (:%s {id: %r}) " % (self.indexed_label,
                     self.vertex_id)
            self.v.append(self.vertex_id)
            self.vertex_id += 1
        self.execute(query)

    def remove_vertex(self):
        vertex_id = random_element(self.v)
        result = self.execute(
            "MATCH (n:%s {id: %r}) OPTIONAL MATCH (n)-[r]-() "
            "DETACH DELETE n RETURN n.id, labels(n), r.id" %
            (self.indexed_label, vertex_id))
        if result:
            process_vertex_ids = set()
            for row in result:
                # remove vertex but note there could be duplicates
                vertex_id = row['n.id']
                if vertex_id not in process_vertex_ids:
                    process_vertex_ids.add(vertex_id)
                    self.v.remove(vertex_id)
                    for label in row['labels(n)']:
                        if (label != self.indexed_label):
                            self.graph.labels[label].remove(vertex_id)
                # remove edge
                edge_id = row['r.id']
                if edge_id != None:
                    self.e.remove(edge_id)

    def create_edge(self):
        creation = self.execute(
            "MATCH (from:%s {id: %r}), (to:%s {id: %r}) "
            "CREATE (from)-[e:EdgeType {id: %r}]->(to) RETURN e" % (
            self.indexed_label, random_element(self.v), self.indexed_label,
            random_element(self.v), self.edge_id))
        if creation:
            self.e.append(self.edge_id)
            self.edge_id += 1

    def remove_edge(self):
        edge_id = random_element(self.e)
        result = self.execute("MATCH (:%s)-[e {id: %r}]->(:%s) DELETE e "
                              "RETURN e.id" % (self.indexed_label, edge_id,
                                               self.indexed_label))
        if result:
            self.e.remove(edge_id)

    def add_label(self):
        vertex_id = random_element(self.v)
        label = random.choice(list(self.graph.labels.keys()))
        # add a label on a vertex that didn't have that label
        # yet (we need that for book-keeping)
        result = self.execute("MATCH (v:%s {id: %r}) WHERE not v:%s SET v:%s "
                              "RETURN v.id" % (self.indexed_label, vertex_id,
                                               label, label))
        if result:
            self.graph.labels[label].append(vertex_id)

    def update_global_vertices(self):
        lo = random.randint(0, self.vertex_id)
        hi = lo + int(self.vertex_id * 0.01)
        num = random.randint(0, 2 ** 20)
        self.execute("MATCH (n) WHERE n.id > %d AND n.id < %d "
                     "SET n.value = %d" % (lo, hi, num))

    def update_global_edges(self):
        lo = random.randint(0, self.edge_id)
        hi = lo + int(self.edge_id * 0.01)
        num = random.randint(0, 2 ** 20)
        self.execute("MATCH ()-[e]->() WHERE e.id > %d AND e.id < %d "
                     "SET e.value = %d" % (lo, hi, num))

    def verify_graph(self):
        """ Checks if the local info corresponds to DB state """

        def test(obj, length, message):
            assert len(obj) == length, message % (len(obj), length)

        def get(query, key):
            ret = self.execute(query)
            assert ret != None, "Query '{}' returned 'None'!".format(query)
            return [row[key] for row in ret]

        test(self.v, get("MATCH (n:{}) RETURN count(n)".format(
             self.indexed_label), "count(n)")[0],
             "Expected %d vertices, found %d")
        test(self.e, get("MATCH (:{0})-[r]->(:{0}) RETURN count(r)".format(
             self.indexed_label), "count(r)")[0],
             "Expected %d edges, found %d")
        for lab, exp in self.graph.labels.items():
            test(exp, get("MATCH (n:%s:%s) RETURN count(n)" % (
                 self.indexed_label, lab), "count(n)")[0],
                 "Expected %d vertices with label '{}', found %d".format(
                 lab))

        log.info("Runner %d graph verification success:", self.sid)
        log.info("\tExecuted %d queries in %.2f seconds",
                 self.executed_queries, time.time() - self._start_time)
        log.info("\tGraph has %d vertices and %d edges",
                 len(self.v), len(self.e))
        for label in sorted(self.graph.labels.keys()):
            log.info("\tVertices with label '%s': %d",
                     label, len(self.graph.labels[label]))
        failures = self.graph.query_failures()
        if failures:
            log.info("\tQuery failed (reason: count)")
            for reason, count in failures.items():
                log.info("\t\t'%s': %d", reason, count)

    def run_loop(self, vertex_batch, query_count, max_time, verify):
        # start the test
        start_time = last_verify = time.time()

        # initial batched vertex creation
        for _ in range(self.graph.vertex_count // vertex_batch):
            if (time.time() - start_time) / 60 > max_time \
                    or self.executed_queries > query_count:
                break
            self.create_vertices(vertex_batch)
        self.create_vertices(self.graph.vertex_count % vertex_batch)

        # run rest
        while self.executed_queries < query_count:
            now_time = time.time()
            if (now_time - start_time) / 60 > max_time:
                break

            if verify > 0 and (now_time - last_verify) > verify:
                self.verify_graph()
                last_verify = now_time

            ratio_e = len(self.e) / self.graph.edge_count
            ratio_v = len(self.v) / self.graph.vertex_count

            # try to edit vertices globally
            if bernoulli(0.01):
                self.update_global_vertices()

            # try to edit edges globally
            if bernoulli(0.01):
                self.update_global_edges()

            # prefer adding/removing edges whenever there is an edge
            # disbalance and there is enough vertices
            if ratio_v > 0.5 and abs(1 - ratio_e) > 0.2:
                if bernoulli(ratio_e / 2.0):
                    self.remove_edge()
                else:
                    self.create_edge()
                continue

            # if we are near vertex balance, we can also do updates
            # instad of update / deletes
            if abs(1 - ratio_v) < 0.5 and bernoulli(0.5):
                self.add_label()
                continue

            if bernoulli(ratio_v / 2.0):
                self.remove_vertex()
            else:
                self.create_vertices(1)


def runner(params):
    num, args = params
    driver = common.argument_driver(args)
    graph = Graph(args.vertex_count // args.worker_count,
                  args.edge_count // args.worker_count)
    log.info("Starting query runner process")
    session = GraphSession(num, graph, driver.session())
    session.run_loop(args.vertex_batch, args.max_queries // args.worker_count,
                     args.max_time, args.verify)
    log.info("Runner %d executed %d queries", num, session.executed_queries)
    driver.close()


def parse_args():
    argp = common.connection_argument_parser()
    argp.add_argument("--logging", default="INFO",
                      choices=["INFO", "DEBUG", "WARNING", "ERROR"],
                      help="Logging level")
    argp.add_argument("--vertex-count", type=int, required=True,
                      help="The average number of vertices in the graph")
    argp.add_argument("--edge-count", type=int, required=True,
                      help="The average number of edges in the graph")
    argp.add_argument("--vertex-batch", type=int, default=200,
                      help="The number of vertices to be created "
                      "simultaneously")
    argp.add_argument("--prop-count", type=int, default=5,
                      help="The max number of properties on a node")
    argp.add_argument("--max-queries", type=int, default=2 ** 30,
                      help="Maximum number of queries  to execute")
    argp.add_argument("--max-time", type=int, default=2 ** 30,
                      help="Maximum execution time in minutes")
    argp.add_argument("--verify", type=int, default=0,
                      help="Interval (seconds) between checking local info")
    argp.add_argument("--worker-count", type=int, default=1,
                      help="The number of workers that operate on the graph "
                      "independently")
    return argp.parse_args()


def main():
    args = parse_args()
    if args.logging:
        logging.basicConfig(level=args.logging)
        logging.getLogger("neo4j").setLevel(logging.WARNING)
    log.info("Starting Memgraph long running test")

    # cleanup and create indexes
    driver = common.argument_driver(args)
    driver.session().run("MATCH (n) DETACH DELETE n").consume()
    for i in range(args.worker_count):
        label = INDEX_FORMAT.format(i)
        driver.session().run("CREATE INDEX ON :%s(id)" % label).consume()
    driver.close()

    params = [(i, args) for i in range(args.worker_count)]
    with multiprocessing.Pool(args.worker_count) as p:
        p.map(runner, params, 1)

    log.info("All query runners done")


if __name__ == '__main__':
    main()

File: tests/stress/long_running.cpp
// Copyright 2022 Memgraph Ltd.
//
// Use of this software is governed by the Business Source License
// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
// License, and you may not use this file except in compliance with the Business Source License.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0, included in the file
// licenses/APL.txt.

#include <atomic>
#include <fstream>
#include <random>
#include <set>
#include <thread>

#include <fmt/format.h>
#include <gflags/gflags.h>

#include "communication/bolt/client.hpp"
#include "io/network/endpoint.hpp"
#include "utils/exceptions.hpp"
#include "utils/timer.hpp"

using EndpointT = memgraph::io::network::Endpoint;
using ClientContextT = memgraph::communication::ClientContext;
using ClientT = memgraph::communication::bolt::Client;
using ValueT = memgraph::communication::bolt::Value;
using QueryDataT = memgraph::communication::bolt::QueryData;
using ExceptionT = memgraph::communication::bolt::ClientQueryException;

DEFINE_string(address, "127.0.0.1", "Server address");
DEFINE_int32(port, 7687, "Server port");
DEFINE_string(username, "", "Username for the database");
DEFINE_string(password, "", "Password for the database");
DEFINE_bool(use_ssl, false, "Set to true to connect with SSL to the server.");

DEFINE_int32(vertex_count, 0, "The average number of vertices in the graph per worker");
DEFINE_int32(edge_count, 0, "The average number of edges in the graph per worker");
DEFINE_int32(prop_count, 5, "The max number of properties on a node");
DEFINE_uint64(max_queries, 1 << 30, "Maximum number of queries to execute");
DEFINE_int32(max_time, 1, "Maximum execution time in minutes");
DEFINE_int32(verify, 0, "Interval (seconds) between checking local info");
DEFINE_int32(worker_count, 1, "The number of workers that operate on the graph independently");
DEFINE_bool(global_queries, true, "If queries that modify globally should be executed sometimes");

DEFINE_string(stats_file, "", "File into which to write statistics.");

/**
 * Encapsulates a Graph and a Bolt session and provides CRUD op functions.
 * Also defines a run-loop for a generic executor, and a graph state
 * verification function.
 */
class GraphSession {
 public:
  GraphSession(int id, std::atomic<uint64_t> *init_complete)
      : id_(id),
        init_complete_(init_complete),
        indexed_label_(fmt::format("indexed_label{}", id)),
        generator_{std::random_device{}()} {
    for (int i = 0; i < FLAGS_prop_count; ++i) {
      auto label = fmt::format("label{}", i);
      labels_.insert(label);
      labels_vertices_.insert({label, {}});
    }

    EndpointT endpoint(FLAGS_address, FLAGS_port);
    client_ = std::make_unique<ClientT>(context_);
    client_->Connect(endpoint, FLAGS_username, FLAGS_password);
  }

 private:
  uint64_t id_;
  std::atomic<uint64_t> *init_complete_;
  ClientContextT context_{FLAGS_use_ssl};
  std::unique_ptr<ClientT> client_;

  uint64_t vertex_id_{0};
  uint64_t edge_id_{0};

  std::set<uint64_t> vertices_;
  std::set<uint64_t> edges_;

  std::string indexed_label_;
  std::set<std::string> labels_;

  std::map<std::string, std::set<uint64_t>> labels_vertices_;

  uint64_t executed_queries_{0};
  std::map<std::string, uint64_t> query_failures_;

  std::mt19937 generator_;

  memgraph::utils::Timer timer_;

 private:
  double GetRandom() { return std::generate_canonical<double, 10>(generator_); }

  bool Bernoulli(double p) { return GetRandom() < p; }

  uint64_t RandomElement(std::set<uint64_t> &data) {
    uint64_t min = *data.begin(), max = *data.rbegin();
    uint64_t val = std::floor(GetRandom() * (max - min) + min);
    auto it = data.lower_bound(val);
    return *it;
  }

  std::string RandomElement(std::set<std::string> &data) {
    uint64_t pos = std::floor(GetRandom() * data.size());
    auto it = data.begin();
    std::advance(it, pos);
    return *it;
  }

  void AddQueryFailure(const std::string &what) {
    auto it = query_failures_.find(what);
    if (it != query_failures_.end()) {
      ++it->second;
    } else {
      query_failures_.insert(std::make_pair(what, 1));
    }
  }

  QueryDataT ExecuteWithoutCatch(const std::string &query) {
    SPDLOG_INFO("Runner {} executing query: {}", id_, query);
    executed_queries_ += 1;
    return client_->Execute(query, {});
  }

  QueryDataT Execute(const std::string &query) {
    try {
      return ExecuteWithoutCatch(query);
    } catch (const ExceptionT &e) {
      AddQueryFailure(e.what());
      return QueryDataT();
    }
  }

  void CreateVertices(uint64_t vertices_count) {
    if (vertices_count == 0) return;
    QueryDataT ret;
    try {
      ret = ExecuteWithoutCatch(fmt::format("UNWIND RANGE({}, {}) AS r CREATE (n:{} {{id: r}}) RETURN count(n)",
                                            vertex_id_, vertex_id_ + vertices_count - 1, indexed_label_));
    } catch (const ExceptionT &e) {
      LOG_FATAL("Runner {} vertices creation failed because of: {}", id_, e.what());
    }
    MG_ASSERT(ret.records.size(), "Runner {} the vertices creation query should have returned a row!", id_);

    MG_ASSERT(ret.records[0][0].ValueInt() == vertices_count, "Runner {} created {} vertices instead of {}!", id_,
              ret.records[0][0].ValueInt(), vertices_count);
    for (uint64_t i = 0; i < vertices_count; ++i) {
      MG_ASSERT(vertices_.insert(vertex_id_ + i).second, "Runner {} vertex with ID {} shouldn't exist!", id_,
                vertex_id_ + i);
    }
    vertex_id_ += vertices_count;
  }

  void RemoveVertex() {
    auto vertex_id = RandomElement(vertices_);
    auto ret =
        Execute(fmt::format("MATCH (n:{} {{id: {}}}) OPTIONAL MATCH (n)-[r]-() "
                            "WITH n, n.id as n_id, labels(n) as labels_n, collect(r.id) as r_ids "
                            "DETACH DELETE n RETURN n_id, labels_n, r_ids",
                            indexed_label_, vertex_id));
    if (ret.records.size() > 0) {
      std::set<uint64_t> processed_vertices;
      auto &record = ret.records[0];
      // remove vertex but note there could be duplicates
      auto n_id = record[0].ValueInt();
      if (processed_vertices.insert(n_id).second) {
        MG_ASSERT(vertices_.erase(n_id), "Runner {} vertex with ID {} should exist!", id_, n_id);
        for (auto &label : record[1].ValueList()) {
          if (label.ValueString() == indexed_label_) {
            continue;
          }
          labels_vertices_[label.ValueString()].erase(n_id);
        }
      }
      // remove edge
      auto &edges = record[2];
      for (auto &edge : edges.ValueList()) {
        if (edge.type() == ValueT::Type::Int) {
          MG_ASSERT(edges_.erase(edge.ValueInt()), "Runner {} edge with ID {} should exist!", id_, edge.ValueInt());
        }
      }
    }
  }

  void CreateEdges(uint64_t edges_count) {
    if (edges_count == 0) return;
    auto edges_per_node = (double)edges_count / vertices_.size();
    MG_ASSERT(std::abs(edges_per_node - (int64_t)edges_per_node) < 0.0001,
              "Runner {} edges per node not a whole number!", id_);

    QueryDataT ret;
    try {
      ret =
          ExecuteWithoutCatch(fmt::format("MATCH (a:{0}) WITH a "
                                          "UNWIND range(0, {1}) AS i WITH a, tointeger(rand() * {2}) AS id "
                                          "MATCH (b:{0} {{id: id}}) WITH a, b "
                                          "CREATE (a)-[e:EdgeType {{id: counter(\"edge\", {3})}}]->(b) "
                                          "RETURN count(e)",
                                          indexed_label_, (int64_t)edges_per_node - 1, vertices_.size(), edge_id_));
    } catch (const ExceptionT &e) {
      LOG_FATAL("Runner {} edges creation failed because of: {}", id_, e.what());
    }
    MG_ASSERT(ret.records.size(), "Runner {} the edges creation query should have returned a row!", id_);

    uint64_t count = ret.records[0][0].ValueInt();
    for (uint64_t i = 0; i < count; ++i) {
      MG_ASSERT(edges_.insert(edge_id_ + i).second, "Runner {} edge with ID {} shouldn't exist!", id_, edge_id_ + i);
    }
    edge_id_ += count;
  }

  void CreateEdge() {
    auto ret = Execute(
        fmt::format("MATCH (from:{} {{id: {}}}), (to:{} {{id: {}}}) "
                    "CREATE (from)-[e:EdgeType {{id: {}}}]->(to) RETURN 1",
                    indexed_label_, RandomElement(vertices_), indexed_label_, RandomElement(vertices_), edge_id_));
    if (ret.records.size() > 0) {
      MG_ASSERT(edges_.insert(edge_id_).second, "Runner {} edge with ID {} shouldn't exist!", id_, edge_id_);
      ++edge_id_;
    }
  }

  void AddLabel() {
    auto vertex_id = RandomElement(vertices_);
    auto label = RandomElement(labels_);
    // add a label on a vertex that didn't have that label
    // yet (we need that for book-keeping)
    auto ret = Execute(fmt::format("MATCH (v:{} {{id: {}}}) WHERE not v:{} SET v:{} RETURN 1", indexed_label_,
                                   vertex_id, label, label));
    if (ret.records.size() > 0) {
      labels_vertices_[label].insert(vertex_id);
    }
  }

  void UpdateGlobalVertices() {
    uint64_t vertex_id = *vertices_.rbegin();
    uint64_t lo = std::floor(GetRandom() * vertex_id);
    uint64_t hi = std::floor(lo + vertex_id * 0.01);
    uint64_t num = std::floor(GetRandom() * (1 << 30));
    Execute(fmt::format("MATCH (n) WHERE n.id > {} AND n.id < {} SET n.value = {}", lo, hi, num));
  }

  void UpdateGlobalEdges() {
    uint64_t vertex_id = *vertices_.rbegin();
    uint64_t lo = std::floor(GetRandom() * vertex_id);
    uint64_t hi = std::floor(lo + vertex_id * 0.01);
    uint64_t num = std::floor(GetRandom() * (1 << 30));
    Execute(fmt::format("MATCH ()-[e]->() WHERE e.id > {} AND e.id < {} SET e.value = {}", lo, hi, num));
  }

  void CheckGraphProjection() {
    uint64_t vertex_id = *vertices_.rbegin();
    uint64_t lo = std::floor(GetRandom() * vertex_id);
    uint64_t hi = std::floor(lo + vertex_id * 0.01);

    Execute(fmt::format(
        "MATCH p=()-[e]->() WHERE e.id > {} AND e.id < {} WITH project(p) as graph WITH graph.nodes as nodes "
        "UNWIND nodes as n RETURN n.x "
        "as x ORDER BY x DESC",
        lo, hi));
    Execute(fmt::format(
        "MATCH p=()-[e]->() WHERE e.id > {} AND e.id < {} WITH project(p) as graph WITH graph.edges as edges "
        "UNWIND edges as e RETURN e.prop as y ORDER BY y DESC",
        lo, hi));
  }

  /** Checks if the local info corresponds to DB state */
  void VerifyGraph() {
    // helper lambda for set verification
    auto test_set = [this](const auto &query, const auto &container, const auto &what) {
      QueryDataT ret;
      try {
        ret = ExecuteWithoutCatch(query);
      } catch (const ExceptionT &e) {
        LOG_FATAL("Runner {} couldn't execute {} ID retrieval because of: {}", id_, what, e.what());
      }
      MG_ASSERT(ret.records.size() == container.size(), "Runner {} expected {} {}, found {}!", id_, container.size(),
                what, ret.records.size());
      for (size_t i = 0; i < ret.records.size(); ++i) {
        MG_ASSERT(ret.records[i].size() == 1, "Runner {} received an invalid ID row for {}!", id_, what);
        auto id = ret.records[i][0].ValueInt();
        MG_ASSERT(container.find(id) != container.end(),
                  "Runner {} couldn't find ID {} for {}! Examined {} items out "
                  "of {} items!",
                  id_, id, what, i, ret.records.size());
      }
    };

    test_set(fmt::format("MATCH (n:{}) RETURN n.id", indexed_label_), vertices_, "vertices");
    test_set(fmt::format("MATCH (:{0})-[r]->(:{0}) RETURN r.id", indexed_label_), edges_, "edges");

    for (auto &item : labels_vertices_) {
      test_set(fmt::format("MATCH (n:{}:{}) RETURN n.id", indexed_label_, item.first), item.second,
               fmt::format("vertices with label '{}'", item.first));
    }

    // generate report
    std::ostringstream report;
    report << fmt::format("Runner {} graph verification success:", id_) << std::endl
           << fmt::format("\tExecuted {} queries in {:.2f} seconds", executed_queries_, timer_.Elapsed().count())
           << std::endl
           << fmt::format("\tGraph has {} vertices and {} edges", vertices_.size(), edges_.size()) << std::endl;
    for (auto &label : labels_) {
      report << fmt::format("\tVertices with label '{}': {}", label, labels_vertices_[label].size()) << std::endl;
    }
    if (query_failures_.size() > 0) {
      report << "\tQuery failed (reason: count)" << std::endl;
      for (auto &item : query_failures_) {
        report << fmt::format("\t\t'{}': {}", item.first, item.second) << std::endl;
      }
    }
    spdlog::info(report.str());
  }

 public:
  void Run() {
    // initial vertex creation
    CreateVertices(FLAGS_vertex_count);

    // initial edge creation
    CreateEdges(FLAGS_edge_count);

    VerifyGraph();
    double last_verify = timer_.Elapsed().count();

    // notify that we completed our initialization
    init_complete_->fetch_add(-1, std::memory_order_acq_rel);

    // run rest
    while (executed_queries_ < FLAGS_max_queries && timer_.Elapsed().count() / 60.0 < FLAGS_max_time) {
      if (FLAGS_verify > 0 && timer_.Elapsed().count() - last_verify > FLAGS_verify) {
        VerifyGraph();
        last_verify = timer_.Elapsed().count();
      }

      double ratio_e = (double)edges_.size() / (double)FLAGS_edge_count;
      double ratio_v = (double)vertices_.size() / (double)FLAGS_vertex_count;

      // try to edit vertices globally if all workers completed their
      // initialization
      if (FLAGS_global_queries && init_complete_->load(std::memory_order_acquire) == 0) {
        if (Bernoulli(0.01)) {
          UpdateGlobalVertices();
        }

        // try to edit edges globally
        if (Bernoulli(0.01)) {
          UpdateGlobalEdges();
        }
      }

      // if we're missing edges (due to vertex detach delete), add some!
      if (Bernoulli(ratio_e < 0.9)) {
        CreateEdge();
        continue;
      }

      // if we are near vertex balance, we can also do updates
      // instad of update / deletes
      if (std::fabs(1.0 - ratio_v) < 0.5 && Bernoulli(0.5)) {
        AddLabel();
        continue;
      }

      if (Bernoulli(ratio_v / 2.0)) {
        RemoveVertex();
      } else {
        CreateVertices(1);
      }
      CheckGraphProjection();
    }

    // final verification
    VerifyGraph();
  }

  uint64_t GetExecutedQueries() { return executed_queries_; }

  uint64_t GetFailedQueries() {
    uint64_t failed = 0;
    for (const auto &item : query_failures_) {
      failed += item.second;
    }
    return failed;
  }
};

int main(int argc, char **argv) {
  gflags::ParseCommandLineFlags(&argc, &argv, true);

  memgraph::communication::SSLInit sslInit;

  MG_ASSERT(FLAGS_vertex_count > 0, "Vertex count must be greater than 0!");
  MG_ASSERT(FLAGS_edge_count > 0, "Edge count must be greater than 0!");

  spdlog::info("Starting Memgraph long running test");

  // create client
  EndpointT endpoint(FLAGS_address, FLAGS_port);
  ClientContextT context(FLAGS_use_ssl);
  ClientT client(context);
  client.Connect(endpoint, FLAGS_username, FLAGS_password);

  // cleanup and create indexes
  client.Execute("MATCH (n) DETACH DELETE n", {});
  for (int i = 0; i < FLAGS_worker_count; ++i) {
    client.Execute(fmt::format("CREATE INDEX ON :indexed_label{}", i), {});
    client.Execute(fmt::format("CREATE INDEX ON :indexed_label{}(id)", i), {});
  }

  // close client
  client.Close();

  // sessions
  std::vector<GraphSession> sessions;
  std::atomic<uint64_t> init_complete(FLAGS_worker_count);
  sessions.reserve(FLAGS_worker_count);
  for (int i = 0; i < FLAGS_worker_count; ++i) {
    sessions.emplace_back(i, &init_complete);
  }

  // workers
  std::vector<std::thread> threads;
  for (int i = 0; i < FLAGS_worker_count; ++i) {
    threads.push_back(std::thread([&, i]() { sessions[i].Run(); }));
  }

  for (int i = 0; i < FLAGS_worker_count; ++i) {
    threads[i].join();
  }

  if (FLAGS_stats_file != "") {
    uint64_t executed = 0, failed = 0;
    for (int i = 0; i < FLAGS_worker_count; ++i) {
      executed += sessions[i].GetExecutedQueries();
      failed += sessions[i].GetFailedQueries();
    }
    std::ofstream stream(FLAGS_stats_file);
    stream << executed << std::endl << failed << std::endl;
    spdlog::info("Written statistics to file: {}", FLAGS_stats_file);
  }

  spdlog::info("All query runners done");

  return 0;
}
